### PR TITLE
GC: Add option to fail a load request instead of closing the container for GC violations

### PR DESCRIPTION
**Pull request by markfields**
The SweepReadyUsageDetection code as-is only can close the container.  When we shared this with some Msft internal partners, they expressed a preference for having the datastore request fail but let the session persist.

This change gives a way to throw the error instead which will get caught by the datastore's request routing code and turned into a 500 response.  So the session will be fine but the component will not load.

### CODE REVIEW

**Comment by markfields** [comment]
@agarwal-navin @tyler-cai-microsoft we may have to patch internal.1.4 to unblock eDiscovery, I'm thinking about putting this in there too, in case Loop wants to simulate Tombstoning now.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:45]: Why are we doing an includes check rather than a comparison check?

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:37]: nit: spacing

**Review by tyler-cai-microsoft — APPROVED** [review]
LGTM

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:45]: So you can indicate multiple modes (e.g. "interactiveClientClose, summarizer") or something.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:37]: I'm not confused, sad actually. But there is no "sad" reaction.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts
Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:45]: I'd prefer 1 flag per feature, it doesn't really matter to me, but it may confuse other developers as this is a new pattern.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts
Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:37]: It was probably the auto correct functionality on your code editor

File: packages/runtime/container-runtime/src/test/gcSweepReadyUsageDetection.spec.ts
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/test/gcSweepReadyUsageDetection.spec.ts:211]: Should the error be renamed to "objectUsedAfterGarbageCollected"? The work "unreferenced" is redundant and does not add anything.

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:150]: The error message should be renamed to something that an end-user / partners can understand. I don't think they care about summarizer or non-summarizer clients. Maybe it should just say "object used after garbage collected"?
We can add another field in the errorDetails to say this was an interactive client.

File: packages/runtime/container-runtime/src/test/gcSweepReadyUsageDetection.spec.ts
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/test/gcSweepReadyUsageDetection.spec.ts:213]: What does the response contain when a data store that is sweep ready is requested? The status should be 500 but what about the message and stack?

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:150]: Sure, good feedback.  clientType should be set anyway on all these logs

File: packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/gcSweepReadyUsageDetection.ts:45]: For the record:  We discussed it as a FC and came to this decision together :P

File: packages/runtime/container-runtime/src/test/gcSweepReadyUsageDetection.spec.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/test/gcSweepReadyUsageDetection.spec.ts:211]: How about more concise "garbageObjectUsedError" - since I updated the error message to "Object used after garbage collected"

File: packages/runtime/container-runtime/src/test/gcSweepReadyUsageDetection.spec.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/test/gcSweepReadyUsageDetection.spec.ts:213]: value is "Error".  Not useful.  I made a fix to that code too... will be sure to test it.

File: packages/runtime/container-runtime/src/test/gcSweepReadyUsageDetection.spec.ts
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/gcSweepReadyUsageDetection.spec.ts]: "garbageObject" sounds funny to me :D I will leave it to you to choose the name (naming is hard!). Another option is "objectUsedAfterGC".

File: packages/runtime/runtime-utils/src/dataStoreHelpers.ts
Comment (agarwal-navin) [inline packages/runtime/runtime-utils/src/dataStoreHelpers.ts:47]: Consider calling `generateErrorWithStack()` here directly so its not called unnecessarily if there is already a stack.

**Review by agarwal-navin — APPROVED** [review]

File: packages/runtime/container-runtime/src/test/gcSweepReadyUsageDetection.spec.ts
Comment (markfields) [reply packages/runtime/container-runtime/src/test/gcSweepReadyUsageDetection.spec.ts]: Haha me too

File: packages/runtime/runtime-utils/src/dataStoreHelpers.ts
Comment (markfields) [reply packages/runtime/runtime-utils/src/dataStoreHelpers.ts:47]: Actually that will change the callstack, it's correct as-is.

**Comment by markfields** [comment]
Abandoning this as a patch on this release branch, rebased to main.  See #12876

**Pull request closed without merging**
